src/bruteforce.rs

```diff
@@ -0,0 +1,647 @@
+use crate::context::{Context, IRule, IStr};
+use crate::forest::{
+    dynamic::{CxAndGrammar, OwnedHandle},
+    Node,
+};
+use crate::input::{Input, InputMatch, Range};
+use crate::parser::{ParseResult, Parser};
+use crate::rule::{Rule, SepKind};
+use cyclotron::bruteforce;
+use std::cell::RefCell;
+use std::cmp::Reverse;
+use std::collections::{BTreeSet, BinaryHeap, HashMap, HashSet, VecDeque};
+use std::fmt;
+use std::hash::Hash;
+use std::iter;
+use std::ops::{RangeFrom, RangeTo};
+use std::rc::Rc;
+
+#[derive(Clone, Default, Debug)]
+struct CachedParse {
+    lengths: Rc<BTreeSet<usize>>,
+    approx_forest: Rc<ApproxForest>,
+}
+
+// HACK(eddyb) hide `approx_forest` from the cyclotron, no need to
+// ensure a fixpoint for it (and `HashMap` could make that tricky).
+impl PartialEq for CachedParse {
+    fn eq(&self, other: &Self) -> bool {
+        self.lengths == other.lengths
+    }
+}
+
+impl Eq for CachedParse {}
+
+/// An approximation of a parse forest, erasing the end of ranges
+/// wherever convenient, to minimize allocations (especially for lists,
+/// where this is linear instead of quadratic).
+///
+/// NB: Requires a second validation pass to produce a proper SPPF.
+#[derive(Default, Debug)]
+struct ApproxForest {
+    possibilities: HashMap<(IRule, RangeFrom<usize>), (Option<RangeTo<usize>>, SmallSet<usize>)>,
+}
+
+#[derive(Debug)]
+enum SmallSet<T> {
+    One(T),
+    Many(BTreeSet<T>),
+}
```

Comment:
Might be cheaper to use e.g. [`im::OrdSet`](https://docs.rs/im/15.0.0/im/struct.OrdSet.html) for this.